File: examples/goblins/hello-node/worker.mjs
import { readFile, writeFile } from "node:fs/promises";

async function readObject(path, label) {
  const value = JSON.parse(await readFile(path, "utf8"));
  if (value === null || Array.isArray(value) || typeof value !== "object") {
    throw new Error(`${label} must be a JSON object`);
  }
  return value;
}

const input = await readObject(process.env.GOBLIN_INPUT_PATH, "input");
await readObject(process.env.GOBLIN_CONTEXT_PATH, "context");

const runId = process.env.GOBLIN_RUN_ID ?? "unknown-run";
const kind = process.env.GOBLIN_KIND ?? "example.hello-node";
const target = input.target ?? "World";

console.log(`Node goblin says hello to ${target}. The event loop bows.`);

const result = {
  status: "success",
  data: {
    message: "Hello World",
    language: "node",
    runtime: "Node.js 22",
    kind,
    run_id: runId,
    target,
    input,
    quote: "The royal event loop never blocks the throne room.",
  },
  artifacts: [],
  metrics: {},
  handoff: [],
  error: null,
};

await writeFile(process.env.GOBLIN_RESULT_PATH, JSON.stringify(result, null, 2));

File: examples/goblins/hello-ruby/worker.rb
require "json"

def read_object(path, label)
  value = JSON.parse(File.read(path))
  raise "#{label} must be a JSON object" unless value.is_a?(Hash)

  value
end

input = read_object(ENV.fetch("GOBLIN_INPUT_PATH"), "input")
read_object(ENV.fetch("GOBLIN_CONTEXT_PATH"), "context")

run_id = ENV.fetch("GOBLIN_RUN_ID", "unknown-run")
kind = ENV.fetch("GOBLIN_KIND", "example.hello-ruby")
target = input.fetch("target", "World")

puts "Ruby goblin says hello to #{target}. The crown likes expressive syntax."

result = {
  status: "success",
  data: {
    message: "Hello World",
    language: "ruby",
    runtime: "Ruby 3.3",
    kind: kind,
    run_id: run_id,
    target: target,
    input: input,
    quote: "A cheerful block can carry a royal decree."
  },
  artifacts: [],
  metrics: {},
  handoff: [],
  error: nil
}

File.write(ENV.fetch("GOBLIN_RESULT_PATH"), JSON.pretty_generate(result))
